File: wrappers/vb/LibRangeMap/ObjectRangeMapper.vb
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq
Imports System.Reflection
Imports System.Text.Json

Public Class ObjectRangeMapper
    Private Shared ReadOnly JsonOptions As New JsonSerializerOptions With {
        .PropertyNamingPolicy = Nothing,
        .WriteIndented = False
    }

    Private Const SpecType As String = "map_range"

    Private ReadOnly _schema As Dictionary(Of String, Object)
    Private ReadOnly _hasMissingValue As Boolean
    Private ReadOnly _missingValue As Object

    Public ReadOnly Property AllowUnknown As Boolean
    Public ReadOnly Property AllowEmpty As Boolean
    Public ReadOnly Property Name As String
    Public ReadOnly Property Schema As IReadOnlyDictionary(Of String, Object)
        Get
            Return _schema
        End Get
    End Property

    Public ReadOnly Property MissingValue As Object
        Get
            Return _missingValue
        End Get
    End Property

    Public ReadOnly Property HasMissingValue As Boolean
        Get
            Return _hasMissingValue
        End Get
    End Property

    Public Sub New(schema As IDictionary(Of String, Object), Optional allowUnknown As Boolean = False, Optional allowEmpty As Boolean = False, Optional name As String = Nothing, Optional hasMissingValue As Boolean = False, Optional missingValue As Object = Nothing)
        If schema Is Nothing Then
            Throw New ArgumentNullException(NameOf(schema), "schema payload must not be null.")
        End If

        Dim normalized = ValidateSchema(schema)
        If Not allowEmpty AndAlso normalized.Count = 0 Then
            Throw New ArgumentException("schema must include at least one field.")
        End If

        If name IsNot Nothing AndAlso name.Length = 0 Then
            Throw New ArgumentException("name must not be empty.", NameOf(name))
        End If

        _schema = normalized
        AllowUnknown = allowUnknown
        AllowEmpty = allowEmpty
        Me.Name = name

        _hasMissingValue = hasMissingValue
        _missingValue = missingValue
    End Sub

    Public Function MapValue(value As Object) As Dictionary(Of String, Object)
        Return Map(value)
    End Function

    Public Function Map(value As Object) As Dictionary(Of String, Object)
        If value Is Nothing Then
            Throw New ArgumentNullException(NameOf(value), "map payload must not be null.")
        End If

        If TypeOf value Is IDictionary Then
            Return MapDictionary(DirectCast(value, IDictionary))
        End If

        Return MapObject(value)
    End Function

    Public Function Spec() As ObjectMapperSpec
        Return ToSpec()
    End Function

    Public Function ToJson() As String
        Return JsonSerializer.Serialize(Spec(), JsonOptions)
    End Function

    Public Shared Function FromJson(json As String) As ObjectRangeMapper
        Dim spec = JsonSerializer.Deserialize(Of ObjectMapperSpec)(json, JsonOptions)
        If spec Is Nothing Then
            Throw New InvalidOperationException("mapper spec JSON produced null.")
        End If

        Return FromSpec(spec)
    End Function

    Public Shared Function FromSpec(spec As ObjectMapperSpec) As ObjectRangeMapper
        If spec.SpecVersion <> CurrentSpecVersion Then
            Throw New ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", NameOf(spec))
        End If

        If spec.MapperType <> SpecType Then
            Throw New ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", NameOf(spec))
        End If

        If spec.Schema Is Nothing Then
            Throw New ArgumentException("schema is required for map mapper.", NameOf(spec))
        End If

        Dim schema As New Dictionary(Of String, Object)(spec.Schema.Count, StringComparer.Ordinal)
        For Each kvp In spec.Schema
            If String.IsNullOrWhiteSpace(kvp.Key) Then
                Throw New ArgumentException("schema field names must be non-empty strings.")
            End If
            schema(kvp.Key) = MapperFromSpecElement(kvp.Value)
        Next

        Dim missing As Object = Nothing
        If spec.HasMissingValue Then
            If spec.MissingValue.HasValue Then
                missing = JsonElementToObject(spec.MissingValue.Value)
            End If
        End If

        Return New ObjectRangeMapper(schema, spec.AllowUnknown, spec.AllowEmpty, spec.Name, spec.HasMissingValue, missing)
    End Function

    Private Shared Function ValidateSchema(schema As IDictionary(Of String, Object)) As Dictionary(Of String, Object)
        Dim normalized As New Dictionary(Of String, Object)(schema.Count, StringComparer.Ordinal)

        For Each field In schema
            If String.IsNullOrWhiteSpace(field.Key) Then
                Throw New InvalidOperationException("schema field names must be non-empty strings.")
            End If
            If field.Value Is Nothing Then
                Throw New ArgumentNullException($"schema field '{field.Key}' mapper")
            End If
            If normalized.ContainsKey(field.Key) Then
                Throw New ArgumentException($"duplicate schema field '{field.Key}'.")
            End If

            normalized(field.Key) = field.Value
        Next

        Return normalized
    End Function

    Private Function MapDictionary(mapInput As IDictionary) As Dictionary(Of String, Object)
        Dim input As New Dictionary(Of String, Object)(StringComparer.Ordinal)
        For Each entry As DictionaryEntry In mapInput
            Dim key = TryCast(entry.Key, String)
            If String.IsNullOrWhiteSpace(key) Then
                Throw New ArgumentException("map keys must be non-empty strings.")
            End If
            input(key) = entry.Value
        Next

        If input.Count = 0 AndAlso Not AllowEmpty Then
            Throw New ArgumentException("map payload must not be empty unless allowEmpty is true.")
        End If

        If Not AllowUnknown Then
            Dim unknownKeys = input.Keys.Where(Function(key) Not _schema.ContainsKey(key)).OrderBy(Function(key) key).ToArray()
            If unknownKeys.Length > 0 Then
                Throw New ArgumentException($"unknown schema fields: {String.Join(", ", unknownKeys)}; set allowUnknown=True to accept extra fields.")
            End If
        End If

        Return MapBySchema(input, "dictionary")
    End Function

    Private Function MapObject(value As Object) As Dictionary(Of String, Object)
        Dim input As New Dictionary(Of String, Object)(StringComparer.Ordinal)
        Dim valueType = value.GetType()

        For Each entry In ReadableMembers(valueType, value)
            input(entry.Key) = entry.Value
        Next

        If input.Count = 0 AndAlso Not AllowEmpty Then
            Throw New ArgumentException("map payload must not be empty unless allowEmpty is true.")
        End If

        If Not AllowUnknown Then
            Dim unknownMembers = ReadableMemberNames(valueType).Where(Function(fieldName) Not _schema.ContainsKey(fieldName)).OrderBy(Function(fieldName) fieldName).ToArray()
            If unknownMembers.Length > 0 Then
                Throw New ArgumentException($"map object has unknown fields: {String.Join(", ", unknownMembers)}; set allowUnknown=True to accept extras.")
            End If
        End If

        Return MapBySchema(input, "object")
    End Function

    Private Function MapBySchema(input As Dictionary(Of String, Object), context As String) As Dictionary(Of String, Object)
        Dim mapped As New Dictionary(Of String, Object)(StringComparer.Ordinal)

        For Each kvp In _schema
            Dim fieldName = kvp.Key
            Dim fieldMapper = kvp.Value
            Dim hasValue As Boolean
            Dim rawValue As Object = Nothing

            hasValue = input.TryGetValue(fieldName, rawValue)
            If hasValue Then
                mapped(fieldName) = MapField(fieldMapper, fieldName, rawValue, context)
                Continue For
            End If

            If HasMissingValue Then
                mapped(fieldName) = MissingValue
                Continue For
            End If

            Throw New KeyNotFoundException($"missing required field '{fieldName}' in {context}.")
        Next

        Return mapped
    End Function

    Private Shared Function MapField(mapper As Object, fieldName As String, fieldValue As Object, context As String) As Object
        If mapper Is Nothing Then
            Throw New InvalidOperationException($"schema field '{fieldName}' has no mapper configured.")
        End If

        If TypeOf mapper Is IntegerRangeMapper Then
            Return DirectCast(mapper, IntegerRangeMapper).MapValue(ConvertToLong(fieldValue, fieldName, context))
        End If

        If TypeOf mapper Is FloatRangeMapper Then
            Return DirectCast(mapper, FloatRangeMapper).MapValue(ConvertToDouble(fieldValue, fieldName, context))
        End If

        If TypeOf mapper Is BooleanRangeMapper Then
            Return DirectCast(mapper, BooleanRangeMapper).MapValue(ConvertToBoolean(fieldValue, fieldName, context))
        End If

        If TypeOf mapper Is BytesRangeMapper Then
            Return DirectCast(mapper, BytesRangeMapper).MapValue(ConvertToBytePayload(fieldValue, fieldName, context))
        End If

        If TypeOf mapper Is CategoricalRangeMapper Then
            Return DirectCast(mapper, CategoricalRangeMapper).MapValue(ConvertToToken(fieldValue, fieldName, context))
        End If

        If TypeOf mapper Is TextRangeMapper Then
            Return DirectCast(mapper, TextRangeMapper).MapValue(ConvertToSingleCharacter(fieldValue, fieldName, context))
        End If

        If TypeOf mapper Is ObjectRangeMapper Then
            Return DirectCast(mapper, ObjectRangeMapper).Map(fieldValue)
        End If

        Dim mapperType = mapper.GetType()
        If mapperType.Name.StartsWith("SequenceRangeMapper") Then
            Return InvokeSequenceMapper(mapper, fieldName, fieldValue, context)
        End If

        Throw New ArgumentException($"schema field '{fieldName}' must provide a supported mapper type.")
    End Function

    Private Shared Function InvokeSequenceMapper(mapper As Object, fieldName As String, fieldValue As Object, context As String) As Object
        If fieldValue Is Nothing Then
            Throw New ArgumentNullException($"field '{fieldName}'")
        End If

        Dim mapMethod As MethodInfo = GetCompatibleMapperMethod(mapper.GetType(), "MapValue", fieldValue)
        If mapMethod Is Nothing Then
            mapMethod = GetCompatibleMapperMethod(mapper.GetType(), "Map", fieldValue)
        End If

        If mapMethod Is Nothing Then
            Throw New ArgumentException($"schema field '{fieldName}' sequence mapper is incompatible for {context}.")
        End If

        Try
            Return mapMethod.Invoke(mapper, New Object() {fieldValue})
        Catch ex As TargetInvocationException When ex.InnerException IsNot Nothing
            Throw ex.InnerException
        End Try
    End Function

    Private Shared Function GetCompatibleMapperMethod(mapperType As Type, methodName As String, fieldValue As Object) As MethodInfo
        Dim methods = mapperType.GetMethods(BindingFlags.Public Or BindingFlags.Instance).Where(Function(method) method.Name = methodName AndAlso method.GetParameters().Length = 1)
        For Each method In methods
            Dim parameterType = method.GetParameters()(0).ParameterType
            If fieldValue Is Nothing Then
                If Not parameterType.IsValueType OrElse Nullable.GetUnderlyingType(parameterType) IsNot Nothing Then
                    Return method
                End If
                Continue For
            End If

            If parameterType Is GetType(Object) OrElse parameterType.IsAssignableFrom(fieldValue.GetType()) Then
                Return method
            End If
        Next

        Return Nothing
    End Function

    Private Shared Function ConvertToLong(value As Object, fieldName As String, context As String) As Long
        If value Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects integer values only when mapping from {context}.")
        End If

        If TypeOf value Is Boolean Then
            Throw New ArgumentException($"field '{fieldName}' expects integer values only.")
        End If

        Try
            Select Case True
                Case TypeOf value Is Byte
                    Return CLng(DirectCast(value, Byte))
                Case TypeOf value Is SByte
                    Return CLng(DirectCast(value, SByte))
                Case TypeOf value Is Short
                    Return CLng(DirectCast(value, Short))
                Case TypeOf value Is UShort
                    Return CLng(DirectCast(value, UShort))
                Case TypeOf value Is Integer
                    Return CLng(DirectCast(value, Integer))
                Case TypeOf value Is UInteger
                    Return CLng(DirectCast(value, UInteger))
                Case TypeOf value Is Long
                    Return CLng(DirectCast(value, Long))
                Case TypeOf value Is ULong
                    Return Convert.ToInt64(DirectCast(value, ULong))
                Case Else
                    Throw New ArgumentException($"field '{fieldName}' expects integer values only.")
            End Select
        Catch ex As OverflowException
            Throw New ArgumentException($"field '{fieldName}' integer value is out of supported range.")
        End Try
    End Function

    Private Shared Function ConvertToDouble(value As Object, fieldName As String, context As String) As Double
        If value Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects float values only when mapping from {context}.")
        End If

        Select Case True
            Case TypeOf value Is Single
                Return CDbl(DirectCast(value, Single))
            Case TypeOf value Is Double
                Return CDbl(DirectCast(value, Double))
            Case TypeOf value Is Decimal
                Return CDbl(DirectCast(value, Decimal))
            Case Else
                Throw New ArgumentException($"field '{fieldName}' expects float values only.")
        End Select
    End Function

    Private Shared Function ConvertToBoolean(value As Object, fieldName As String, context As String) As Boolean
        If value Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects boolean values only when mapping from {context}.")
        End If

        If TypeOf value Is Boolean Then
            Return DirectCast(value, Boolean)
        End If

        Throw New ArgumentException($"field '{fieldName}' expects boolean values only.")
    End Function

    Private Shared Function ConvertToBytePayload(value As Object, fieldName As String, context As String) As Byte()
        If value Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects a byte array when mapping from {context}.")
        End If

        Dim bytes = TryCast(value, Byte())
        If bytes Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects a byte array when mapping from {context}.")
        End If

        Return bytes
    End Function

    Private Shared Function ConvertToToken(value As Object, fieldName As String, context As String) As String
        If value Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects a categorical token when mapping from {context}.")
        End If

        Dim token = TryCast(value, String)
        If token Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects a categorical token when mapping from {context}.")
        End If

        Return token
    End Function

    Private Shared Function ConvertToSingleCharacter(value As Object, fieldName As String, context As String) As Char
        If value Is Nothing Then
            Throw New ArgumentException($"field '{fieldName}' expects a single character when mapping from {context}.")
        End If

        If TypeOf value Is Char Then
            Return DirectCast(value, Char)
        End If

        Dim text = TryCast(value, String)
        If text IsNot Nothing AndAlso text.Length = 1 Then
            Return text(0)
        End If

        Throw New ArgumentException($"field '{fieldName}' expects a single character when mapping from {context}.")
    End Function

    Private Shared Function ReadableMemberNames(valueType As Type) As String()
        Dim names As New List(Of String)()

        For Each propertyInfo In valueType.GetProperties(BindingFlags.Instance Or BindingFlags.Public)
            If propertyInfo.GetMethod IsNot Nothing AndAlso propertyInfo.GetIndexParameters().Length = 0 Then
                names.Add(propertyInfo.Name)
            End If
        Next

        For Each fieldInfo In valueType.GetFields(BindingFlags.Instance Or BindingFlags.Public)
            names.Add(fieldInfo.Name)
        Next

        Return names.ToArray()
    End Function

    Private Shared Function ReadableMembers(valueType As Type, value As Object) As Dictionary(Of String, Object)
        Dim members As New Dictionary(Of String, Object)(StringComparer.Ordinal)

        For Each propertyInfo In valueType.GetProperties(BindingFlags.Instance Or BindingFlags.Public)
            If propertyInfo.GetMethod Is Nothing OrElse propertyInfo.GetIndexParameters().Length > 0 Then
                Continue For
            End If

            members(propertyInfo.Name) = propertyInfo.GetValue(value)
        Next

        For Each fieldInfo In valueType.GetFields(BindingFlags.Instance Or BindingFlags.Public)
            members(fieldInfo.Name) = fieldInfo.GetValue(value)
        Next

        Return members
    End Function

    Private Function ToSpec() As ObjectMapperSpec
        Dim schema As New Dictionary(Of String, JsonElement)(StringComparer.Ordinal)

        For Each kvp In _schema
            schema(kvp.Key) = JsonDocument.Parse(SerializeMapper(kvp.Value)).RootElement.Clone()
        Next

        Dim missingElement As JsonElement? = Nothing
        If _hasMissingValue Then
            If _missingValue Is Nothing Then
                Using nullDoc As JsonDocument = JsonDocument.Parse("null")
                    missingElement = nullDoc.RootElement.Clone()
                End Using
            Else
                Using valueDoc As JsonDocument = JsonSerializer.SerializeToDocument(_missingValue, JsonOptions)
                    missingElement = valueDoc.RootElement.Clone()
                End Using
            End If
        End If

        Return New ObjectMapperSpec With {
            .SpecVersion = CurrentSpecVersion,
            .MapperType = SpecType,
            .Schema = schema,
            .AllowUnknown = AllowUnknown,
            .AllowEmpty = AllowEmpty,
            .HasMissingValue = _hasMissingValue,
            .MissingValue = missingElement,
            .Name = Name
        }
    End Function

    Private Shared Function SerializeMapper(mapper As Object) As String
        If TypeOf mapper Is IntegerRangeMapper Then
            Return DirectCast(mapper, IntegerRangeMapper).ToJson()
        End If
        If TypeOf mapper Is FloatRangeMapper Then
            Return DirectCast(mapper, FloatRangeMapper).ToJson()
        End If
        If TypeOf mapper Is BooleanRangeMapper Then
            Return DirectCast(mapper, BooleanRangeMapper).ToJson()
        End If
        If TypeOf mapper Is BytesRangeMapper Then
            Return DirectCast(mapper, BytesRangeMapper).ToJson()
        End If
        If TypeOf mapper Is CategoricalRangeMapper Then
            Return DirectCast(mapper, CategoricalRangeMapper).ToJson()
        End If
        If TypeOf mapper Is TextRangeMapper Then
            Return DirectCast(mapper, TextRangeMapper).ToJson()
        End If
        If TypeOf mapper Is ObjectRangeMapper Then
            Return DirectCast(mapper, ObjectRangeMapper).ToJson()
        End If

        Dim mapperType = mapper.GetType()
        If mapperType.Name.StartsWith("SequenceRangeMapper") Then
            Dim toJsonMethod = mapperType.GetMethod("ToJson")
            If toJsonMethod IsNot Nothing Then
                Return CStr(toJsonMethod.Invoke(mapper, Array.Empty(Of Object)()))
            End If
        End If

        Throw New InvalidOperationException("unsupported mapper type for schema.")
    End Function

    Private Shared Function MapperFromSpecElement(mapperElement As JsonElement) As Object
        If mapperElement.ValueKind <> JsonValueKind.Object Then
            Throw New InvalidOperationException("schema values must be mapper spec objects.")
        End If

        Dim mapperTypeElement As JsonElement
        If Not mapperElement.TryGetProperty("mapper_type", mapperTypeElement) Then
            Throw New InvalidOperationException("schema values require a mapper_type field.")
        End If

        If mapperTypeElement.ValueKind <> JsonValueKind.String Then
            Throw New InvalidOperationException("schema values require a valid mapper_type string.")
        End If

        Dim mapperType = mapperTypeElement.GetString()
        Select Case mapperType
            Case "integer_range"
                Return IntegerRangeMapper.FromJson(mapperElement.GetRawText())
            Case "float_range"
                Return FloatRangeMapper.FromJson(mapperElement.GetRawText())
            Case "boolean_range"
                Return BooleanRangeMapper.FromJson(mapperElement.GetRawText())
            Case "bytes_range"
                Return BytesRangeMapper.FromJson(mapperElement.GetRawText())
            Case "categorical_range"
                Return CategoricalRangeMapper.FromJson(mapperElement.GetRawText())
            Case "text_range"
                Return TextRangeMapper.FromJson(mapperElement.GetRawText())
            Case "map_range"
                Return FromSpec(JsonSerializer.Deserialize(Of ObjectMapperSpec)(mapperElement.GetRawText(), JsonOptions))
            Case "sequence_range"
                Throw New NotSupportedException("sequence_range specs are not supported by ObjectRangeMapper.FromSpec in Visual Basic wrapper yet.")
            Case Else
                Throw New InvalidOperationException($"unsupported mapper_type '{mapperType}'.")
        End Select
    End Function

    Private Shared Function JsonElementToObject(element As JsonElement) As Object
        Select Case element.ValueKind
            Case JsonValueKind.String
                Return element.GetString()
            Case JsonValueKind.Number
                Dim intValue As Long
                If element.TryGetInt64(intValue) Then
                    Return intValue
                End If
                Return element.GetDouble()
            Case JsonValueKind.True
                Return True
            Case JsonValueKind.False
                Return False
            Case JsonValueKind.Null
                Return Nothing
            Case JsonValueKind.Object
                Return JsonSerializer.Deserialize(Of Dictionary(Of String, Object))(element.GetRawText(), JsonOptions)
            Case JsonValueKind.Array
                Return JsonSerializer.Deserialize(Of List(Of Object))(element.GetRawText(), JsonOptions)
            Case Else
                Throw New InvalidOperationException("unsupported missing_value JSON kind.")
        End Select
    End Function
End Class

File: wrappers/csharp/LibRangeMap/ObjectRangeMapper.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LibRangeMap;

public sealed class ObjectRangeMapper
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        WriteIndented = false
    };

    private const string SpecVersion = MapperSpec.CurrentSpecVersion;
    private const string SpecType = "map_range";

    private readonly Dictionary<string, object> _schema;
    private readonly object? _missingValue;
    private readonly bool _hasMissingValue;

    public bool AllowUnknown { get; }
    public bool AllowEmpty { get; }
    public string? Name { get; }

    public ObjectRangeMapper(
        IDictionary<string, object> schema,
        bool allowUnknown = false,
        bool allowEmpty = false,
        string? name = null,
        bool hasMissingValue = false,
        object? missingValue = null)
    {
        if (schema is null)
        {
            throw new ArgumentNullException(nameof(schema), "schema payload must not be null.");
        }

        _schema = ValidateSchema(schema);
        if (!allowEmpty && _schema.Count == 0)
        {
            throw new ArgumentException("schema must include at least one field.", nameof(schema));
        }

        if (name is not null && name.Length == 0)
        {
            throw new ArgumentException("name must not be empty.", nameof(name));
        }

        if (hasMissingValue)
        {
            _missingValue = missingValue;
        }

        AllowUnknown = allowUnknown;
        AllowEmpty = allowEmpty;
        Name = name;
        _hasMissingValue = hasMissingValue;
    }

    public IReadOnlyDictionary<string, object> Schema => _schema;

    public object? MissingValue => _missingValue;

    public bool HasMissingValue => _hasMissingValue;

    public Dictionary<string, object?> Map(object value)
    {
        if (value is null)
        {
            throw new ArgumentNullException(nameof(value), "map payload must not be null.");
        }

        if (value is IDictionary dict)
        {
            return MapDictionary(dict);
        }

        return MapObject(value);
    }

    public string ToJson()
    {
        return JsonSerializer.Serialize(ToSpec(), JsonOptions);
    }

    public static ObjectRangeMapper FromJson(string json)
    {
        var spec = JsonSerializer.Deserialize<ObjectMapperSpec>(json, JsonOptions)
            ?? throw new InvalidOperationException("mapper spec JSON produced null.");
        return FromSpec(spec);
    }

    public static ObjectRangeMapper FromSpec(ObjectMapperSpec spec)
    {
        if (spec.SpecVersion != SpecVersion)
        {
            throw new ArgumentException($"unsupported spec_version '{spec.SpecVersion}'.", nameof(spec));
        }

        if (spec.MapperType != SpecType)
        {
            throw new ArgumentException($"unsupported mapper_type '{spec.MapperType}'.", nameof(spec));
        }

        if (spec.Schema is null)
        {
            throw new ArgumentException("schema is required for map mapper.", nameof(spec));
        }

        var schema = new Dictionary<string, object>(spec.Schema.Count, StringComparer.Ordinal);
        foreach (var kvp in spec.Schema)
        {
            if (string.IsNullOrWhiteSpace(kvp.Key))
            {
                throw new ArgumentException("schema field names must be non-empty.");
            }

            schema[kvp.Key] = MapperFromSpecElement(kvp.Value);
        }

        var hasMissingValue = spec.HasMissingValue;
        object? missingValue = null;
        if (hasMissingValue)
        {
            if (spec.MissingValue is JsonElement element)
            {
                missingValue = JsonElementToObject(element);
            }
        }

        return new ObjectRangeMapper(
            schema,
            spec.AllowUnknown,
            spec.AllowEmpty,
            spec.Name,
            hasMissingValue,
            missingValue);
    }

    private Dictionary<string, object?> MapDictionary(IDictionary dict)
    {
        var mapped = new Dictionary<string, object?>(StringComparer.Ordinal);
        var input = new Dictionary<string, object?>(StringComparer.Ordinal);

        foreach (DictionaryEntry item in dict)
        {
            if (item.Key is not string key || key.Length == 0)
            {
                throw new InvalidOperationException("map keys must be non-empty strings.");
            }

            input[key] = item.Value;
        }

        if (input.Count == 0 && !AllowEmpty)
        {
            throw new ArgumentException("map payload must not be empty unless allow_empty is true.");
        }

        var unknownKeys = input.Keys.Where(k => !_schema.ContainsKey(k)).ToArray();
        if (!AllowUnknown && unknownKeys.Length > 0)
        {
            throw new ArgumentException(
                $"unknown schema fields: {string.Join(\", \", unknownKeys)}; set allowUnknown=true to ignore extra fields.");
        }

        foreach (var (fieldName, mapper) in _schema)
        {
            if (input.TryGetValue(fieldName, out var fieldValue))
            {
                mapped[fieldName] = MapField(mapper, fieldName, fieldValue);
            }
            else if (_hasMissingValue)
            {
                mapped[fieldName] = _missingValue;
            }
            else
            {
                throw new KeyNotFoundException($"missing required field '{fieldName}' in map input.");
            }
        }

        return mapped;
    }

    private Dictionary<string, object?> MapObject(object value)
    {
        var type = value.GetType();
        var mapped = new Dictionary<string, object?>(StringComparer.Ordinal);
        var foundAtLeastOne = false;
        if (!AllowUnknown)
        {
            var memberNames = ReadableMemberNames(type);
            var unexpected = memberNames.Where(fieldName => !_schema.ContainsKey(fieldName)).ToArray();
            if (unexpected.Length > 0)
            {
                throw new ArgumentException(
                    $"map object has unknown fields: {string.Join(\", \", unexpected)}; set allowUnknown=true to accept extras.");
            }
        }

        foreach (var (fieldName, mapper) in _schema)
        {
            if (TryGetObjectMember(value, type, fieldName, out var fieldValue))
            {
                foundAtLeastOne = true;
                mapped[fieldName] = MapField(mapper, fieldName, fieldValue);
            }
            else if (_hasMissingValue)
            {
                mapped[fieldName] = _missingValue;
            }
            else
            {
                throw new KeyNotFoundException($"missing required field '{fieldName}' in object input.");
            }
        }

        if (!foundAtLeastOne && !AllowEmpty)
        {
            throw new ArgumentException("map payload must not be empty unless allow_empty is true.");
        }

        return mapped;
    }

    private static IEnumerable<string> ReadableMemberNames(Type type)
    {
        foreach (var property in type.GetProperties(BindingFlags.Instance | BindingFlags.Public))
        {
            if (property.GetMethod is not null && property.GetIndexParameters().Length == 0)
            {
                yield return property.Name;
            }
        }

        foreach (var field in type.GetFields(BindingFlags.Instance | BindingFlags.Public))
        {
            yield return field.Name;
        }
    }

    private static bool TryGetObjectMember(object value, Type type, string fieldName, out object? fieldValue)
    {
        var property = type.GetProperty(fieldName, BindingFlags.Instance | BindingFlags.Public);
        if (property is not null)
        {
            if (property.GetMethod is null)
            {
                fieldValue = null;
                return false;
            }

            fieldValue = property.GetValue(value);
            return true;
        }

        var field = type.GetField(fieldName, BindingFlags.Instance | BindingFlags.Public);
        if (field is not null)
        {
            fieldValue = field.GetValue(value);
            return true;
        }

        fieldValue = null;
        return false;
    }

    private object MapField(object mapper, string fieldName, object? value)
    {
        return mapper switch
        {
            IntegerRangeMapper integerMapper => integerMapper.MapValue(ConvertToLong(value, fieldName)),
            FloatRangeMapper floatMapper => floatMapper.MapValue(ConvertToDouble(value, fieldName)),
            BooleanRangeMapper booleanMapper => booleanMapper.MapValue(ConvertToBoolean(value, fieldName)),
            BytesRangeMapper bytesMapper => bytesMapper.MapValue(ConvertToByte(value, fieldName)),
            CategoricalRangeMapper categoricalMapper => categoricalMapper.MapValue(value),
            TextRangeMapper textMapper => textMapper.MapValue(ConvertToChar(value, fieldName)),
            SequenceRangeMapper sequenceMapper => sequenceMapper.Map(value),
            ObjectRangeMapper objectMapper => objectMapper.Map(value),
            _ => throw new ArgumentException(
                $"schema field '{fieldName}' must provide a supported mapper type with map methods.")
        };
    }

    private static long ConvertToLong(object? value, string fieldName)
    {
        return value switch
        {
            sbyte v => v,
            byte v => v,
            short v => v,
            ushort v => v,
            int v => v,
            uint v => checked((long)v),
            long v => v,
            ulong v => checked((long)v),
            _ => throw new ArgumentException($"field '{fieldName}' expects integer values only.")
        };
    }

    private static double ConvertToDouble(object? value, string fieldName)
    {
        return value switch
        {
            float v => v,
            double v => v,
            decimal v => (double)v,
            _ => throw new ArgumentException($"field '{fieldName}' expects float values only.")
        };
    }

    private static bool ConvertToBoolean(object? value, string fieldName)
    {
        return value is bool v
            ? v
            : throw new ArgumentException($"field '{fieldName}' expects boolean values only.");
    }

    private static long ConvertToByte(object? value, string fieldName)
    {
        return value is byte v
            ? v
            : throw new ArgumentException($"field '{fieldName}' expects byte values only.");
    }

    private static char ConvertToChar(object? value, string fieldName)
    {
        if (value is char v)
        {
            return v;
        }

        if (value is string text && text.Length == 1)
        {
            return text[0];
        }

        throw new ArgumentException($"field '{fieldName}' expects a single character.");
    }

    private static object MapperFromSpecElement(JsonElement specElement)
    {
        if (specElement.ValueKind != JsonValueKind.Object)
        {
            throw new InvalidOperationException("schema values must be mapper spec objects.");
        }

        if (!specElement.TryGetProperty("mapper_type", out var mapperTypeElement) ||
            mapperTypeElement.ValueKind != JsonValueKind.String)
        {
            throw new InvalidOperationException("schema values require a mapper_type.");
        }

        var mapperType = mapperTypeElement.GetString();

        return mapperType switch
        {
            "integer_range" => IntegerRangeMapper.FromJson(specElement.GetRawText()),
            "float_range" => FloatRangeMapper.FromJson(specElement.GetRawText()),
            "boolean_range" => BooleanRangeMapper.FromJson(specElement.GetRawText()),
            "bytes_range" => BytesRangeMapper.FromJson(specElement.GetRawText()),
            "categorical_range" => CategoricalRangeMapper.FromJson(specElement.GetRawText()),
            "sequence_range" => SequenceRangeMapper.FromSpec(
                JsonSerializer.Deserialize<SequenceMapperSpec>(specElement.GetRawText(), JsonOptions)
                    ?? throw new InvalidOperationException("mapper spec JSON produced null.")
            ),
            "text_range" => TextRangeMapper.FromJson(specElement.GetRawText()),
            "map_range" => FromSpec(
                JsonSerializer.Deserialize<ObjectMapperSpec>(specElement.GetRawText(), JsonOptions)
                    ?? throw new InvalidOperationException("mapper spec JSON produced null.")
            ),
            _ => throw new InvalidOperationException($"unsupported mapper_type '{mapperType}'.")
        };
    }

    private static object? JsonElementToObject(JsonElement element)
    {
        return element.ValueKind switch
        {
            JsonValueKind.String => element.GetString(),
            JsonValueKind.Number => element.TryGetInt64(out var i64) ? i64 : element.GetDouble(),
            JsonValueKind.True => true,
            JsonValueKind.False => false,
            JsonValueKind.Null => null,
            JsonValueKind.Object => JsonSerializer.Deserialize<Dictionary<string, object>>(element.GetRawText(), JsonOptions),
            JsonValueKind.Array => JsonSerializer.Deserialize<List<object>>(element.GetRawText(), JsonOptions),
            _ => throw new InvalidOperationException("unsupported missing_value JSON kind.")
        };
    }

    private static Dictionary<string, object> ValidateSchema(IDictionary<string, object> schema)
    {
        var normalized = new Dictionary<string, object>(schema.Count, StringComparer.Ordinal);
        foreach (var kvp in schema)
        {
            if (string.IsNullOrWhiteSpace(kvp.Key))
            {
                throw new InvalidOperationException("schema field names must be non-empty strings.");
            }

            if (kvp.Value is null)
            {
                throw new ArgumentNullException($"schema field '{kvp.Key}' mapper");
            }

            normalized[kvp.Key] = kvp.Value;
        }

        return normalized;
    }

    private ObjectMapperSpec ToSpec()
    {
        var schema = new Dictionary<string, JsonElement>(_schema.Count);
        foreach (var (fieldName, mapper) in _schema)
        {
            var mappedSpec = mapper switch
            {
                IntegerRangeMapper integerMapper => integerMapper.ToJson(),
                FloatRangeMapper floatMapper => floatMapper.ToJson(),
                BooleanRangeMapper booleanMapper => booleanMapper.ToJson(),
                BytesRangeMapper bytesMapper => bytesMapper.ToJson(),
                CategoricalRangeMapper categoricalMapper => categoricalMapper.ToJson(),
                SequenceRangeMapper sequenceMapper => sequenceMapper.ToJson(),
                TextRangeMapper textMapper => textMapper.ToJson(),
                ObjectRangeMapper objectMapper => objectMapper.ToJson(),
                _ => throw new InvalidOperationException($"unsupported mapper type for field '{fieldName}'.")
            };

            schema[fieldName] = JsonDocument.Parse(mappedSpec).RootElement.Clone();
        }

        JsonElement? missingValueElement = null;
        if (_hasMissingValue)
        {
            if (_missingValue is null)
            {
                using var doc = JsonDocument.Parse("null");
                missingValueElement = doc.RootElement.Clone();
            }
            else
            {
                using var doc = JsonSerializer.SerializeToDocument(_missingValue, JsonOptions);
                missingValueElement = doc.RootElement.Clone();
            }
        }

        return new ObjectMapperSpec
        {
            SpecVersion = SpecVersion,
            MapperType = SpecType,
            Schema = schema,
            AllowUnknown = AllowUnknown,
            AllowEmpty = AllowEmpty,
            HasMissingValue = _hasMissingValue,
            MissingValue = missingValueElement,
            Name = Name
        };
    }
}

public sealed class ObjectMapperSpec
{
    [JsonPropertyName("spec_version")]
    public string SpecVersion { get; init; } = MapperSpec.CurrentSpecVersion;

    [JsonPropertyName("mapper_type")]
    public string MapperType { get; init; } = "map_range";

    [JsonPropertyName("schema")]
    public Dictionary<string, JsonElement>? Schema { get; init; }

    [JsonPropertyName("allow_unknown")]
    public bool AllowUnknown { get; init; }

    [JsonPropertyName("allow_empty")]
    public bool AllowEmpty { get; init; }

    [JsonPropertyName("has_missing_value")]
    public bool HasMissingValue { get; init; }

    [JsonPropertyName("missing_value")]
    public JsonElement? MissingValue { get; init; }

    [JsonPropertyName("name")]
    public string? Name { get; init; }
}
